pkg/common/commands/resetoffset/controller/dataplane.go

```diff
@@ -0,0 +1,213 @@
+func (r *Reconciler) stopConsumerGroups(ctx context.Context, resetOffset *kafkav1alpha1.ResetOffset, services map[string]ctrl.Service, refInfo *refmappers.RefInfo) error {
+	return r.sendConsumerGroupAsyncCommands(ctx, resetOffset, services, refInfo, commands.StopConsumerGroupOpCode)
+}
+
+// sendConsumerGroupAsyncCommands sends ConsumerGroupAsyncCommands to the specified control-protocol Services in
+// parallel and blocks waiting for all the AsyncCommandResults.  Any errors are returned in a single multi-error.
+func (r *Reconciler) sendConsumerGroupAsyncCommands(ctx context.Context,
+	resetOffset *kafkav1alpha1.ResetOffset,
+	services map[string]ctrl.Service,
+	refInfo *refmappers.RefInfo,
+	opCode ctrl.OpCode) error {
+
+	// Send To All The ConsumerGroups In Parallel
+	waitGroup := &sync.WaitGroup{}
+	waitGroup.Add(len(services))
+	errChan := make(chan error, len(services))
+	for podIP, service := range services {
+		go func(podIP string, service ctrl.Service) {
+			defer waitGroup.Done()
+			err := r.sendConsumerGroupAsyncCommand(ctx, resetOffset, podIP, service, refInfo, opCode)
+			if err != nil {
+				errChan <- err
+			}
+		}(podIP, service)
+	}
+
+	// Wait For All The AsyncCommand Results
+	waitGroup.Wait()
+
+	// Close & Drain The Error Channel
+	close(errChan)
+	var multiErr error
+	for err := range errChan {
+		multierr.AppendInto(&multiErr, err)
+	}
+
+	// Return Any Errors
+	return multiErr
+}
+
+// sendConsumerGroupAsyncCommand sends a ConsumerGroupAsyncCommand with the specified opCode to the
+// specified pods and blocks waiting for AsyncCommandResult response which is then returned.
+func (r *Reconciler) sendConsumerGroupAsyncCommand(ctx context.Context,
+	resetOffset *kafkav1alpha1.ResetOffset,
+	podIP string,
+	service ctrl.Service,
+	refInfo *refmappers.RefInfo,
+	opCode ctrl.OpCode) error {
+
+	// Get The Logger From Context
+	logger := logging.FromContext(ctx).Desugar().With(zap.String("PodIP", podIP), zap.Int("OpCode", int(opCode)))
+
+	// Generate A CommandID For The ResetOffset
+	commandId, err := generateCommandId(resetOffset, podIP, opCode)
+	if err != nil {
+		logger.Error("Failed to generate Command ID for ResetOffset", zap.Error(err))
+		return fmt.Errorf("failed to generate Command ID for ResetOffset: %v", err)
+	}
+
+	// Create And Send A The ConsumerGroupAsyncCommand to The Control-Protocol Service
+	consumerGroupAsyncCommand := commands.NewConsumerGroupAsyncCommand(commandId, refInfo.TopicName, refInfo.GroupId)
+	err = service.SendAndWaitForAck(opCode, consumerGroupAsyncCommand)
+	if err != nil {
+		logger.Error("Failed to send ConsumerGroup AsyncCommand", zap.Int64("CommandID", commandId), zap.Error(err))
+		return fmt.Errorf("failed to send ConsumerGroup AsyncCommand '%d' : %v", commandId, err)
+	}
+
+	// Wait For The AsyncCommand Result & Return Results
+	return r.waitForAsyncCommandResult(resetOffset, podIP, consumerGroupAsyncCommand)
```

Comment:
So you're making the command send asynchronous blocking the control loop right?

Reply:
Yes, we have made the Stop/Start ConsumerGroupAsyncCommands asynchronous by nature as they are also intended to be used for the Pause/Resume feature.  For the ResetOffset use case though I want to make all the async calls in parallel and then block on their completion before proceeding.